Update outdated output in examples, add a draft document for working with job hierarchies

grondo opened the pull request:
This PR freshens up the output in some existing examples (old-style decimal jobids in output mostly), makes a small change to the index (move CORAL, CORAL2 down the list a bit), adds `domainrefs.py` from flux-core so we can more easily cross-reference flux-core and linux man pages, and finally adds a rough draft of a short document _Working with Flux Job Hierarchies_.

The purpose of the new document is to introduce our more recent tools and improvements to display, query, and connect to jobs in a simple Flux instance hierarchy. The document uses examples to illustrate the tools, and thus it is more like an informal tutorial.

I struggled with the order in which to introduce the concepts. I ended up introducing `flux-proxy` last, but perhaps it should actually be first. I'm open to other suggestions. The main difficulty was collecting all the example output from the same session so that the jobids are consistent. I had a section about high-level vs _native_ URIs as well, but removed it since it largely duplicated text in the `flux-uri(1)` man page and seemed out of place for the shorter examples format. Plus, the hope is now that user's can be largely oblivious to the native Flux connector URIs in the long term.

garlick commented on quickstart.rst:
```diff
@@ -254,15 +310,17 @@ The most basic functionality of these service modules can be tested with the ``f
 
 .. code-block:: console
 
   $ flux ping --rank=3 --count=2 kvs
   3!kvs.ping pad=0 seq=0 time=1.888 ms (CBF78!09552!0!1!3!BBC94)
   3!kvs.ping pad=0 seq=1 time=1.792 ms (CBF78!09552!0!1!3!BBC94)
```
Output is a little different these days:
```
$ flux ping --rank=3 --count=2 kvs
3!kvs.ping pad=0 seq=0 time=0.638 ms (8acf1ea0!08c30389!b07f6e60!27effb20!24ecbd0d!4aa819a3)
3!kvs.ping pad=0 seq=1 time=0.678 ms (8acf1ea0!08c30389!b07f6e60!27effb20!24ecbd0d!4aa819a3)
```

grondo commented:
I'm not sure why `make check` is generating the following error in CI:

```
Warning, treated as error:
unknown directive or role name: core:man1
```

I don't get this error locally, and the RTD build succeeded in building the pages properly. 
Perhaps a missing dependency. I'll be pushing some test builds to find out.

garlick commented on hierarchies.rst:
```diff
@@ -0,0 +1,409 @@
+       ƒcgdiQP grondo   sleep       R      1      1   8.444s fluke108
+  
+
+The depth to which to recurse can be specified via the ``-L, --level=N``
```
could "to which" be dropped without harming the flow of this sentence?

garlick reviewed (approved):
Really nice addition!

I have a few editorial comments on first read through, but they are all pretty minor.  I think the order is fine.

grondo commented:
Ok, I think I found the problem. The workflow was using the latest version of sphinx (4.0.0?) which for some reason chokes on the added `core:manX` domains for the `spelling` check target. I recalled that we pinned the version of sphinx in flux-core to 3.4.3, so I've done that here as well.

grondo commented on quickstart.rst:
```diff
@@ -153,7 +153,10 @@ Starting a Flux Instance
 In order to use Flux, you first must initiate a Flux *instance* or *session*.
 
-A Flux session is composed of a hierarchy of ``flux-broker`` processes which are launched via any parallel launch utility that supports PMI. For example, ``srun``, ``mpiexec.hydra``, etc., or locally for testing via the ``flux start`` command.
+A Flux session is composed of a hierarchy of ``flux-broker`` processes
+which are launched via any parallel launch utility that supports PMI. For
+example, ``srun``, ``mpiexec.hydra``, etc., or locally for testing via the
+``flux start`` command with the ``-s, --test-size=N`` option.
```
Sounds good! I should have mentioned I wrapped all lines in quickstart to ~78 as well, so it may look like more was changed that actually was in this PR.

I dislike 1 line per paragraph in text documents under version control since `diff` is mostly line based, so I thought I'd fix that now.

grondo commented on quickstart.rst:
```diff
@@ -172,30 +178,72 @@ A flux session can be also be started under `Slurm <https://github.com/chaos/slu
+After broker wire up is completed, the Flux session starts an “initial
+program” on rank 0 broker. By default, the initial program is an
+interactive shell, but an alternate program can be supplied on the ``flux
+start`` command line. Once the initial program terminates, the Flux session
```
It seemed like we currently use session and instance interchangeably in the docs now. Want me to just update all uses of session to instance?

garlick replied:
Yeah, probably.   I think we are done with the term "comms session" which is the origin of that I think.

grondo commented on quickstart.rst:
```diff
@@ -254,15 +310,17 @@ The most basic functionality of these service modules can be tested with the ``f
   kvs.ping pad=0 seq=1 time=0.666 ms (1F18F!09552!0!EEE45)
 
-By default the local (or closest) instance of the service is targeted, but a specific rank can be selected with the ``--rank`` option.
+By default the local (or closest) instance of the service is targeted,
+but a specific rank can be selected with the ``--rank`` option.
```
Should I then also drop the statement about using `flux ping` to test the latency to any rank? (It does seem outside the scope of a quickstart, but then `flux ping` itself probably falls outside that scope as well)

garlick replied:
That makes sense - it will keep it focused.

grondo replied:
Ok, will do.

grondo commented:
I've made a pass at addressing PR comments in a couple add-on commits and pushed the result.

Once everything looks good I'll squash and force push.

grondo commented:
BTW, I'm also considering adding `make linkcheck` to the CI. I ran it and it found some outdated links. Any thoughts? One drawback is if external sites change links (as was the case here in a couple instances) it will break unrelated PRs.

garlick commented:
> BTW, I'm also considering adding make linkcheck to the CI

I think it's a good idea.

garlick commented:
Just had another quick skim and LGTM!  Thanks for taking time!

grondo commented:
Ok, I squashed all the commits which addressed comments, and I've experimentally added `make linkcheck` to ci. (We'll see how that goes)

grondo commented:
Looks like `make linkcheck` is working. I've added it to branch protections by default, which should prevent PRs with bad external links from being merged.

This PR will have to be manually merged since it updates Github Actions.

grondo commented:
I actually requested a review from @ryanday36, since it will be great to have his perspective on the new document. The rendered result can be viewed [here](https://flux-framework--124.org.readthedocs.build/en/124/hierarchies.html).

ryanday36 reviewed (approved):
This looks good to me. I really like these tools for viewing / interacting with Flux instance hierarchies.

grondo commented:
Good idea @chu11

On Tue, Jan 18, 2022, 2:22 PM Al Chu ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In hierarchies.rst
> <https://github.com/flux-framework/flux-docs/pull/124#discussion_r787179552>
> :
>
> > +A Flux instance records its depth in the current job hierarchy in the
> +``instance-level`` attribute. This attribute can be useful when descending
> +a hierarchy of jobs. In this guide, the following value of ``PS1`` will
> +be used to indicate the size and depth of the current enclosing instance
> +for reference purposes:
>
> would it be worthwhile to mention that depth begins at 0? It's obvious to
> me, but I'll bet there are "it starts at 1" folks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-docs/pull/124#pullrequestreview-856015653>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFVEUXJCN54GV5B7JO33TLUWXR2BANCNFSM5MH3YDAA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

chu11 commented on hierarchies.rst:
```diff
@@ -0,0 +1,409 @@
+Instance level (depth)
+----------------------
+
+A Flux instance records its depth in the current job hierarchy in the
+``instance-level`` attribute. This attribute can be useful when descending
+a hierarchy of jobs. In this guide, the following value of ``PS1`` will
+be used to indicate the size and depth of the current enclosing instance
+for reference purposes:
```
something like "the initial instance, such as the system instance or you test instance, has an initial depth of 0".

grondo replied:
Thanks for the good suggestion @chu11. I slightly modified the example so that it shows the prompt after executing `flux start --test-size=1`, then added a note about `depth = 0` as suggested.

grondo commented:
I set MWP thinking this would need to be manually merged since `.github/workflows` was modified. But mergify seems to have gone ahead and merged. Thanks for the feedback everyone!
